Allow to change the log's 'colors' parameter

suzuki-shunsuke opened the pull request:
This change keeps the backward compatibility.

Usage:

```javascript
const logger = require('yeoman-environment/util/lib/log');

Adapter.prototype.log = logger({
  colors: {
    skip: 'yellow',
    force: 'yellow',
    create: 'green',
    // Add the custom status 'merge'
    merge: 'green',
    invoke: 'bold',
    conflict: 'red',
    // Change identical's color to green
    identical: 'green',
    info: 'gray'
});
```

suzuki-shunsuke commented:
I'm sorry but I can't understand why the ci has failed.
I can't reproduce the error in the local.
It seems the error has nothing to do with this PR.

SBoudrias commented on lib/util/log.js:
```diff
@@ -76,6 +67,19 @@ module.exports = () => {
 
   _.extend(log, EventEmitter.prototype);
 
+  params = Object.assign({
```
I have a concern here, we do not validate that the user will end up providing all of the colors, so on line 147, we can end up with an undefined value.

suzuki-shunsuke replied:
Thank you for reviewing.
Please see the line 121.

https://github.com/yeoman/environment/pull/93/files#diff-cf49fe9112169caa62101be9cf031893R121

The variable `status` is a key of `params.color`, so `params.colors[status]` is not undefined.

SBoudrias replied:
Oh I see, still this means some log status might not be available to some adapters. I think we should also extend the colors status object to make sure all status are defined; this would also allow to only override some colors:

```js
Adapter.prototype.log = logger({
  colors: {
    force: 'pink',
  }
});
```

suzuki-shunsuke replied:
I see. I have updated PR.

* By default, the logger inherits the default colors
* If `params.ignoreDefaultColors` is `true`, the logger ignores the default colors

Required status depends on other components such as Conflicter.
Some conflicters may not require the default status.
So originally I overrode the default colors intentionally.
But in many cases it is convenient for the logger to inherit the default colors.
So I have added the above changes.

suzuki-shunsuke replied:
I have an idea about the validation of colors.
If we want to validate colors, we should add the `validateLogger` method to the Conflicter.

For example,

```javascript
class Conflicter {
  constructor(adapter, force, actions) {
    this.force = force === true;
    this.adapter = adapter;
    this.conflicts = [];
    this._actions = actions || ['create', 'force', 'conflict', 'identical'];
    if (!this.validateLogger()) {
      // some error handling
    }
  }
  validateLogger() {
    const logger = this.adapter.log;
    return logger && this._actions.all(s => {
      return typeof logger[s] === 'function';
    });
  }
}
```

SBoudrias commented on lib/util/log.js:
```diff
@@ -44,7 +33,41 @@ function formatter(msg, ctx) {
+  info: 'gray'
+});
+
+const initParams = params => {
```
Let's use an extend strategy here, something similar to this:

```js
Object.assign(
    defaultParams,
    overwrites,
    {
        colors: Object.assign(defaultParams.colors, overwrites.colors)
    }
)
```

The current code is a bit too complicated

SBoudrias commented:
Tests are failing due to a linting issue. In this case, the code is using unnecessary `else` clauses.

suzuki-shunsuke commented:
I'm sorry but I forgot to execute the test after the commit https://github.com/yeoman/environment/pull/93/commits/d2872f5da986b6d78329aebfb4b9b92d88018ab4.
I have fixed the lint error and make `initParams()` simple.

suzuki-shunsuke commented:
I have fixed lint errors, but the CI is stil failing in the another reason.
Unfortunately, I can't reproduce the failing.

This is the result when I execute the test at local.

```
❯ node --version
v4.8.7

~/repos/src/github.com/suzuki-shunsuke/environment allow-to-change-log-colors
❯ npm t && echo success

> yeoman-environment@2.0.5 pretest /Users/shunsuke/repos/src/github.com/suzuki-shunsuke/environment
> xo


  lib/util/log.js:130:3
  ⚠  130:3  Unexpected todo comment.  no-warning-comments

  1 warning

> yeoman-environment@2.0.5 test /Users/shunsuke/repos/src/github.com/suzuki-shunsuke/environment
> nyc mocha

Skipping tests for global generators. Please setup `NODE_PATH` environment variable to run it.


  TerminalAdapter
    #prompt()
      ✓ pass its arguments to inquirer
    #diff()

removed added

varlet

      ✓ returns properly colored diffs
    #log()
      ✓ calls console.error and perform strings interpolation
      ✓ substitutes strings correctly when context argument is falsey
      ✓ boolean values
      ✓ #write() numbers
      ✓ #write() objects
    #log
dummy      ✓ #write() pass strings as they are
A number: 1, a string: bla      ✓ #write() accepts util#format style arguments
dummy
      ✓ #writeln() adds a \n at the end
✔ dummy
      ✓ #ok() adds a green "✔ " at the beginning and \n at the end
✖ dummy
      ✓ #error() adds a green "✗ " at the beginning and \n at the end
      statuses
        - #skip()
        - #force()
        - #create()
        - #invoke()
        - #conflict()
        - #identical()
        - #info()

  Environment
    ✓ is an instance of EventEmitter
    constructor
      ✓ take arguments option
      ✓ take arguments parameter option as string
      ✓ take options parameter
      ✓ instantiates a TerminalAdapter if none provided
      ✓ uses the provided object as adapter if any
      ✓ instantiates a mem-fs instance
      ✓ takes multi-word arguments
    #help()
      ✓ output the general help
      ✓ output the help with a custom bin name
    #create()
      ✓ instantiate a generator (158ms)
      ✓ instantiate a scoped generator
      ✓ pass options.arguments
      ✓ pass options.arguments as string
      ✓ pass options.args (as `arguments` alias)
      ✓ prefer options.arguments over options.args
      ✓ default arguments to `env.arguments`
      ✓ pass options.options
      ✓ default options to `env.options` content
      ✓ throws if Generator is not registered
      ✓ add the env as property on the options
      ✓ add the Generator resolved path on the options
      ✓ adds the namespace on the options
      ✓ adds the namespace as called on the options
    #run()
      ✓ runs a registered generator
      ✓ pass args and options to the runned generator
      ✓ without options, it default to env.options
      ✓ without args, it default to env.arguments
      ✓ can take string as args
      ✓ can take no arguments
      ✓ launch error if generator is not found
      ✓ returns the generator
      ✓ correctly append scope in generator hint
    #registerModulePath()
      ✓ resolves to a directory if no file type specified
    #register()
      ✓ store registered generators
      ✓ determine registered Generator namespace and resolved path
      ✓ throw when String is not passed as first parameter
    #registerStub()
      ✓ register a function under a namespace
      ✓ throws if invalid generator
      ✓ throws if invalid namespace
    #namespaces()
      ✓ get the list of namespaces
    #getGeneratorsMeta()
      ✓ get the registered Generators metadatas
    #getGeneratorNames
      ✓ get the registered generators names
    #namespace()
      ✓ create namespace from path
      ✓ create namespace from scoped path
      ✓ handle relative paths
      ✓ handles weird paths
      ✓ works with Windows' paths
      ✓ remove lookups from namespace
      ✓ remove path before the generator name
      ✓ handle paths when multiples lookups are in it
    #get()
      ✓ get a specific generator
      ✓ remove paths from namespace at resolution (for backward compatibility)
      ✓ works with Windows' absolute paths
      ✓ fallback to requiring generator from a file path
      ✓ returns undefined if namespace is not found
    #error()
      ✓ delegate error handling to the listener
      ✓ throws error if no listener is set
      ✓ returns the error
    #alias()
      ✓ apply regex and replace with alternative value
      ✓ apply multiple regex
      ✓ apply latest aliases first
      ✓ alias empty namespace to `:app` by default
    .enforceUpdate()
      ✓ add an adapter
      ✓ add a runLoop
      ✓ add a shared mem-fs instance
    .createEnv()
      ✓ create an environment
    .namespaceToName()
      ✓ convert a namespace to a name

  Environment Resolver
    #lookup()
      ✓ register local generators
      ✓ register generators in scoped packages
      ✓ register non-dependency local generator (699ms)
      ✓ local generators prioritized over global
      - register global generators
      ✓ register symlinked generators
      when there's ancestor node_modules/ folder
        ✓ register generators in ancestor node_modules directory
        ✓ local generators are prioritized over ancestor
    #getNpmPaths()
      with NODE_PATH
        ✓ walk up the CWD lookups dir
        ✓ append NODE_PATH
      without NODE_PATH
        ✓ walk up the CWD lookups dir
        ✓ append best bet if NODE_PATH is unset
        ✓ append default NPM dir depending on your OS
      with NVM_PATH
        ✓ walk up the CWD lookups dir
        ✓ append NVM_PATH
      without NVM_PATH
        ✓ walk up the CWD lookups dir
        ✓ append best bet if NVM_PATH is unset

  Store
    #add() / #get()
      storing as module
        ✓ store and return the module
        ✓ assign meta data to the module
        ✓ assign dummy resolved value (can't determine the path of an instantiated)
      storing as module path
        ✓ store and returns the required module
        ✓ assign meta data to the module
    #namespaces()
      ✓ return stored module namespaces

  util
    ✓ is exposed on the Environment object
    .duplicateEnv()
      ✓ copy environment


  104 passing (38s)
  8 pending

-----------------|----------|----------|----------|----------|----------------|
File             |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------|----------|----------|----------|----------|----------------|
All files        |    92.04 |    79.33 |    91.94 |    91.94 |                |
 lib             |    95.56 |    82.31 |      100 |    95.51 |                |
  adapter.js     |    95.83 |    66.67 |      100 |    95.65 |             67 |
  environment.js |     95.6 |    84.04 |      100 |    95.54 |... 448,475,517 |
  resolver.js    |    94.12 |    76.92 |      100 |    94.12 | 68,100,113,115 |
  store.js       |      100 |      100 |      100 |      100 |                |
 lib/util        |    78.26 |       60 |    66.67 |    77.94 |                |
  log.js         |    75.81 |       60 |    64.29 |    75.41 |... 179,180,183 |
  util.js        |      100 |      100 |      100 |      100 |                |
-----------------|----------|----------|----------|----------|----------------|
success
```

SBoudrias commented on lib/util/log.js:
```diff
@@ -44,7 +33,26 @@ function formatter(msg, ctx) {
+
+const initParams = params => {
+  params = params || {};
+  return params.ignoreDefaultColors ?
```
I don't think we really need this parameters. We should just always extends the default colors to ensure the API compatibility.

suzuki-shunsuke replied:
OK. I have removed it in https://github.com/yeoman/environment/pull/93/commits/5fca3d8a919d82f8be542516effa9a9805272c98.

suzuki-shunsuke commented:
Still there are several points We can modify.

1. remove the function getDefaultColors
2. remove the function initParams
3. extend not "params" but only "params.colors"

I think we don't have to modify these points, but I want to know your opinions.

suzuki-shunsuke commented:
For example, we can modify as following.

```javascript
// replace getDefaultColors to a variable
const defaultColors = {
  skip: 'yellow',
  force: 'yellow',
  create: 'green',
  invoke: 'bold',
  conflict: 'red',
  identical: 'cyan',
  info: 'gray'
};
```


```javascript
module.exports = params => {
  // remove initParams
  // extend not params but only params.colors
  const colors = Object.assign({}, defaultColors, params.colors || {});
```

SBoudrias commented:
Yeah, the functions aren't really necessary, but I think it's fine.

Thanks for the PR!

(looks like a test is broken since npm 5, will need to dig up what's the issue later)

suzuki-shunsuke commented:
@SBoudrias 
Thank you for merging the PR!
I really appreciate your reviewing.
